zyte_common_items/util.py

```diff
@@ -54,3 +60,15 @@ def get_origin(tp) -> Tuple:
     typing.get_origin(tp) is only available startingo on 3.8.
     """
     return getattr(tp, "__origin__", ())
+
+
+def is_url_obj(url: Union[str, _Url]) -> TypeGuard[_Url]:
+    return True if isinstance(url, _Url) else False
+
+
+def url_to_str(url: Union[str, _Url]) -> str:
+    if is_url_obj(url):
+        return str(url)
+    elif isinstance(url, str):
+        return url
+    raise ValueError(f"{url!r} is neither a string nor an instance of RequestURL or ResponseURL.")
```

Comment:
What do you think about simplifying it to something like this?

```suggestion
def url_to_str(url: Union[str, _Url]) -> str:
    if not isinstance(url, (str, _Url)):
        raise ValueError(f"{url!r} is neither a string nor an instance of RequestURL or ResponseURL.")
    return str(url)
```

`is_url_obj` doesn't look required; `_Url` is visible in url_to_str signature, so I'm not sure an extra abstraction helps. 

Alternatively, it can also be simplified by dropping is_url_obj, and checking `isinstance(url, _Url)` directly.

Even if util.py is considered an internal module, it's not underscored, and is_url_obj is not underscored, so it becomes a part of public API; it doesn't seem required.

Reply:
Good idea. The `TypeGuard` approach was an attempt to help mypy discern the return type of the converter via type narrowing. But it's not being used right now because of the upstream issues between mypy and attr's converters.